xmind/tests/test_workbook_saver.py

```diff
@@ -0,0 +1,135 @@
+import logging
+from . import base
+from unittest.mock import patch, MagicMock, mock_open
+from xmind.core.saver import WorkbookSaver
+from xmind.core.const import CONTENT_XML, XMIND_EXT
+
+
+class WorkbookSaverTest(base.Base):
+
+    def setUp(self):
+        super(WorkbookSaverTest, self).setUp()
+        self._workbook = MagicMock()
+        self._test_el = WorkbookSaver(self._workbook)
+
+    def getLogger(self):
+        if not getattr(self, '_logger', None):
+            self._logger = logging.getLogger('WoorkbookSaverTest')
+        return self._logger
+
+    def test_excessive_parameters(self):
+        # _workbook = MagicMock()
+        # _element = WorkbookSaver(_workbook)
+        _parameters = [
+            ('_get_content', 0),
+          #  ('save', 1)
+        ]
+
+        with self.assertRaises(TypeError) as _ex:
+            self._test_el.save(12,13)
+
+        self.getLogger().debug("Got an exception: '%s'", _ex.exception.args[0])
+        self.assertTrue(_ex.exception.args[0].find(
+            "%s() takes" % 'save') != -1)
+
+        for pair in _parameters:
+            with self.subTest(pair=pair):
+                self._test_method_by_excessive_parameters(pair, self._test_el)
+
+    def test_WoorkbookSaver_inint(self):
+        # Call without parameter:
+        with self.assertRaises(TypeError) as _ex1:
+            WorkbookSaver()
+        self.assertTrue(_ex1.exception.args[0].find(
+                "__init__() missing") != -1)
+
+        # Call with excessive parameters
+        with self.assertRaises(TypeError) as _ex2:
+            WorkbookSaver(1,2)
+        self.assertTrue(_ex2.exception.args[0].find(
+                "__init__() takes") != -1)
+
+        # Call with one parameter
+        _test_object = WorkbookSaver('workbook')
+        self.assertEqual(_test_object._workbook,'workbook')
+
+    def test_get_content(self):
+
+        with patch('xmind.utils.join_path') as _utils_join_path:
+            _utils_join_path.return_value = 'joined path'
+            with patch('xmind.utils.temp_dir') as _utils_temp_dir:
+                _utils_temp_dir.return_value = 'temp dir'
+                _m = mock_open()  # to test with statement
+                with patch('codecs.open', _m, create=True):
+                    with patch.object(self._workbook, 'output') as _output:
+                        _el = self._test_el._get_content()
+
+        _utils_temp_dir.assert_called_once_with()
+        _utils_join_path.assert_called_once_with('temp dir', CONTENT_XML)
+        _m.assert_called_once_with('joined path', 'w', encoding="utf-8")
+        _output.assert_called_once_with(_m())
+        self.assertEqual(_el, 'joined path')
+
+    def test_save_with_no_path_no_get_path(self):
+        with patch.object(self._workbook, 'get_path') as _get_path:
```

Comment:
what is the reason to patch ```_workbook```? it's already ```MagicMock``` you can just define
```self._workbook.get_path.return_value = None```
in your test

Reply:
Done